program/c/src/oracle/model/price_model.c

```diff
@@ -1,99 +1,108 @@
 #include "price_model.h"
 #include "../util/avg.h" /* For avg_2_int64 */
 
-#define SORT_NAME  int64_sort_ascending
-#define SORT_KEY_T int64_t
-#include "../sort/tmpl/sort_stable.c"
+/*
+ * Quick select implementation highly optimized for reducing the number of compute units in BPF.
+ *
+ * On Floyd and Rivest's SELECT algorithm" by Krzysztof C. Kiwiel article
+ * mentions that the average case of this algorithm is is 2.5n + o(n) comparisons.
+ * The article itself presents a different algorithm (Floyd and Rivest) which
+ * performs faster in larger arrays but has a higher overhead. Other determinstic
+ * algorithms like median of medians have a higher overhead (or comibnations of
+ * them and quick select) and are not suitable for BPF.
+ */
+int64_t quick_select( int64_t * a, uint64_t n, uint64_t k ) {
+    while (1) {
+        if (n == 1) return a[0];
+
+        int64_t pivot;
+
+        /*
+         * Select median of 3 as pivot which massively reduces the number of
+         * comparisons in the average case and also helps in reducing the likelihood
+         * of worst case time complexity.
+         */
+        {
+            if (a[0] < a[n / 2]) {
+                if (a[n / 2] < a[n - 1]) pivot = a[n / 2];
+                else if (a[0] < a[n - 1]) pivot = a[n - 1];
+                else pivot = a[0];
+            } else {
+                if (a[n / 2] > a[n - 1]) pivot = a[n / 2];
+                else if (a[0] > a[n - 1]) pivot = a[n - 1];
+                else pivot = a[0];
+            }
+        }
+
```

Comment:
The median of 3 as pivot already reduces the number of comparisons for the rest of the algorithm, but I think this can be reduced even further by initially picking the middle point optimistically:

```
        uint64_t mid = n / 2;
        int64_t pivot = a[mid];
        if ((a[0] < pivot) != (a[0] < a[n - 1])) {
            pivot = a[0] < a[n - 1] ? a[n - 1] : a[0];
        }
```

Here, we choose the midpoint optimistically then the condition checks if `a[0]` lies between `mid`  and `a[n-i]`, if the check fails we already have the right choice, if not, then the ternary condition selects between `a[0]` and `a[n-1]` which is the same as the logic you already had.

Not writing this as a suggestion to change as I'm not sure if there's any benefit in this change in terms of reducing instruction count.

Reply:
Never mind there's no benefit, your explicit checks are clearer. Ignore this.